### PR TITLE
RSE-1090: Fix Move/Copy Activity

### PR DESCRIPTION
**Pull request by deb1990**
## Overview
When a user only has permission to a single case type category, the Move/Copy to case popup does not show any cases, when typed.

## Before
before

## After
after

## Technical Details
1. The case type categories for which user has "basic case information" permission, is exposed to frontend using `caseTypeCategoriesWhereUserCanAccessBasicCaseInformation`. Also the same values are exposed within angular context using `case-type-category.provider`.

2. Changed the parameters in `move-copy-activity-action.html`. Previously as `case_type_id.case_type_category` parameter was not used, the api was not returning any values because of permission. 
So now `case_type_id.case_type_category` parameter is set to the first case type category for which the user has "basic case information" permission, this bypasses the permission issue.

Also `'case_id.case_type_id.case_type_category'` is set, using which case type categories from which user can see cases, are loaded.

### Backstop Update
Previously the backstop related npm packages was added to the original `package.json`. But backstop is not required to be installed all the time, to avoid this, all the backstop related packages are moved to a new `package.json` under the `tests/backstop_data` folder.

## Backend Permission Issue Fix
We were having some problems with some functions like [this](https://github.com/compucorp/uk.co.compucorp.civicase/blob/6d50b1a40555f77de3086a0e5b7e3a7db7324ec6/CRM/Civicase/Helper/CaseCategory.php#L213) and [this](https://github.com/compucorp/uk.co.compucorp.civicase/blob/6d50b1a40555f77de3086a0e5b7e3a7db7324ec6/CRM/Civicase/Helper/CaseCategory.php#L241), returning the main case category as one of the categories that the user has access to even though the use does not have access to any case related permission.

The issue is because of the way the case type category permissions work: If for example a contact's permission for `access my cases and activities` is being checked, if the user does not have the permission but if the page is associated to any case category, we try to find the equivalent case category for that permission and the access is granted for that user using the `hook_permission_check`

This works fine in most cases but however in some rare cases, we don't need this equivalent permission logic to happen especially if we really want to know if the contact has real access to a particular case permission and not it's equivalent. To achieve this, we skip running the hook responsible for the equivalent permission logic when the contact ID is passed when running the check. Civicrm would usually not pass the contact ID parameter and would run the permission check for the logged in user so this is a way for us to distinguish between whether to run the hook or not.

### CODE REVIEW

File: CRM/Civicase/Hook/PermissionCheck/CaseCategory.php [hunk 1]
```diff
@@ -123,8 +123,9 @@ private function modifyPermissionCheckForCategory($permission, &$granted, $caseC
    */
   private function shouldRun($permission, $granted) {
     $defaultCasePermissions = array_column($this->caseCategoryPermission->get(), 'name');
+    $isNotBasicCase = $permission != 'basic case information';
 
-    return in_array($permission, $defaultCasePermissions) && !$granted;
+    return in_array($permission, $defaultCasePermissions) && !$granted && $isNotBasicCase;
```
Comment (reneolivo) [inline CRM/Civicase/Hook/PermissionCheck/CaseCategory.php:128]: can we ask @tunbola to review this bit? I'm not familiar with this class.

File: ang/civicase-base/providers/case-type-category.provider.js [hunk 2]
```diff
@@ -10,6 +10,8 @@
     var allCaseTypeCategories = civicaseBaseSettings.caseTypeCategories;
     var caseTypeCategoriesWhereUserCanAccessActivities =
     civicaseBaseSettings.caseTypeCategoriesWhereUserCanAccessActivities;
+    var caseTypeCategoriesWhereUserCanAccessBasicCaseInformation =
+    civicaseBaseSettings.caseTypeCategoriesWhereUserCanAccessBasicCaseInformation;
```
Comment (reneolivo) [inline ang/civicase-base/providers/case-type-category.provider.js:14]: can we add spaces to this statement and the one above?

```js
var name =
  value;
```

File: ang/civicase/activity/actions/services/move-copy-activity-action.html [hunk 3]
```diff
@@ -3,7 +3,10 @@
     <div class="form-group">
       <label class="col-sm-4 control-label">{{ model.ts('Select Case') }}</label>
       <div class="col-sm-8">
-        <input class="form-control" crm-entityref="{entity: 'Case'}" ng-model="model.case_id" />
+        <input class="form-control" crm-entityref="{entity: 'Case', api: {
+            'case_type_id.case_type_category': model.caseTypeCategoriesWithAccessToBasicCaseInformation[0],
+            params: {'case_id.case_type_id.case_type_category': {'IN': model.caseTypeCategoryIDsWithAccessToBasicCaseInformation }}
+          }}" ng-model="model.case_id" />
```
Comment (reneolivo) [inline ang/civicase/activity/actions/services/move-copy-activity-action.html:9]: is it possible to define the API call parameters in the model? This block is a bit hard to follow. We could also test this value if we move it to the model.

File: CRM/Civicase/Hook/PermissionCheck/CaseCategory.php (hunk 1 above)
Comment (deb1990) [reply CRM/Civicase/Hook/PermissionCheck/CaseCategory.php]: This was added by him only :)

File: ang/civicase/activity/actions/services/move-copy-activity-action.service.js [hunk 4]
```diff
@@ -63,6 +65,20 @@
       });
     }
 
+    /**
+     * @returns {object} api parameters for Case.getlist
+     */
+    function getCaseListApiParams () {
+      return {
+        'case_type_id.case_type_category': _.keys(CaseTypeCategory.getCategoriesWithAccessToBasicCaseInformation())[0],
```
Comment (reneolivo) [inline ang/civicase/activity/actions/services/move-copy-activity-action.service.js:73]: let's move this part to a variable, it's too long:

```js
var firstCaseTypeCategoryId =
  _.keys(CaseTypeCategory.getCategoriesWithAccessToBasicCaseInformation())[0];
var allCaseTypeCategoryNames =
  _.values(CaseTypeCategory.getCategoriesWithAccessToBasicCaseInformation());

return {
  'case_type_id.case_type_category': firstCaseTypeCategoryId,
  params: {
    'case_id.case_type_id.case_type_category': {
      IN: allCaseTypeCategoryNames
    }
  }
}
```

Comment (reneolivo) [reply ang/civicase/activity/actions/services/move-copy-activity-action.service.js:73]: Also, a big question. I know it was in the description, but why do we send the first case category ID plus the full list of case category names? Can you please explain a bit more about that?
